### ISSUE TITLE
App::Parse: Check for input / output images

### ISSUE THREAD

**Issue by Lestropie** [MRtrix3/mrtrix3#327]
`lib/app.cpp:495`:

```
      // check for the existence of all specified input files (including optional ones that have been provided)
      // if necessary, also check for pre-existence of any output files with known paths
      //   (if the output is e.g. given as a prefix, the argument should be flagged as type_text())
      for (std::vector<ParsedArgument>::const_iterator i = argument.begin(); i < argument.end(); i++) {
        if ((i->arg->type == ArgFileIn) && !Path::exists (std::string(*i)))
          throw Exception ("required input file \"" + str(*i) + "\" not found");
        if (i->arg->type == ArgFileOut)
          check_overwrite (std::string(*i));
      }
      for (std::vector<ParsedOption>::const_iterator i = option.begin(); i != option.end(); ++i) {
        for (size_t j = 0; j != i->opt->size(); ++j) {
          const Argument& arg = i->opt->operator [](j);
          const char* const name = i->args[j];
          if ((arg.type == ArgFileIn) && !Path::exists (name))
            throw Exception ("input file \"" + str(name) + "\" not found (required for option \"-" + std::string(i->opt->id) + "\")");
          if (arg.type == ArgFileOut)
           check_overwrite (name);
        }
      }
```

This section doesn't check for the presence of input / output images, only general files. Presumably this is because special-case formats (e.g. exporting to multiple images using `[]` notation) won't be correctly handled. However it would be nice if these outlier cases could be detected here, so that in the vast majority of cases, the app back-end will check for the presence of images before running.

DICOM inport should work fine (`Path::exists()` returns true for a directory). Anything with `[]` will need to be skipped, as will piped temp files. Are there any other cases that need to be handled?

**Comment by jdtournier** [comment]
Yes, there's a few other cases. The problem is, they're format-specific, which is why we don't currently check for that - we leave it to the format handler to deal with. Examples include hdr/img pairs for Analyze/NIfTI, mih/dat pairs for mih (although there is _no_ requirement that the data file have the .dat suffix, properly checking would require opening the mih header to parse the `file:` entry), and a few similar issues for obscure outdated formats that we could drop support for. And there's every chance that future formats would also require special handling of this nature... 

Really the only easy to deal with this cleanly while maintaining some level of independence between the image format handling code and the command-line parsing code is to have these checks handled by the image format handling code somehow. The problem is to do it right for image creation, you might need to know quite a bit more about the image you're about to write (especially for the multi-file bracketed syntax) to figure out which actual files are doing to be written. We could handle the open() situation by explicitly loading the images at argument parse time, but since opening input images of typically the first thing that happens, there's not much to be gained there. The image creation problem is trickier because of the prior knowledge that might be required. 

Personally, I think it's much easier to make sure output images are opened early on, before any processing proper happens - that way everything is handled correctly, and you get early termination of the command as desired. It preserves one independence between the image format handling code and the command-line parsing code (which I'm very keen to maintain), and in practice there shouldn't be many commands where this is a problem - I'd rather focus on fixing those as needed...

**Comment by Lestropie** [comment]
Stop sounding so spammy, you didn't make it past my spam filter :-P

Was thinking along the lines of specifying additional virtual handler functions that would 'claim' the format, then check for existing files (within reason). This would also require relevant functions to be written for new formats. E.g.:

* `.mih`: For read, if it's there, do a quick key/value parse to find the data file path, and check that also. For write, also check the corresponding .dat path for an existing file.
* `[].nii`: Check for just the '0' index file for read. For write, query the target directory contents and warn if there are _any_ files there with the specified prefix and the .nii suffix (don't know yet how many will be written).
* `.hdr` & `.img`: Pretty easy.
* `-`: Do nothing.

Early output image open may not always be possible / easy. E.g. for an optional output image it's much easier to detect the option and branch at output time, rather than creating it early and storing a pointer; even more so if properties of the output image header aren't known until processing is complete. Currently in `tckmap` the output image isn't created until the stack is fully unwound (which could be changed, just an example). At least this mechanism would be command-independent, but agreed it might be more work.

Not pushing it, just throwing it out there. Anyone else got any thoughts?
